src/chain/shared_state.rs

```diff
@@ -220,3 +229,66 @@ impl NonEmptyList<(SectionInfo, ProofSet)> {
         }
     }
 }
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct SectionProofBlock {
+    key: BlsPublicKey,
+    sig: BlsSignature,
+}
+
+impl SectionProofBlock {
+    pub fn from_sec_info_with_proofs(sec_info: &SectionInfo, proofs: ProofSet) -> Self {
+        let key = BlsPublicKey::from_section_info(sec_info);
+        let sig = BlsSignature::from_proof_set(proofs);
+        SectionProofBlock { key, sig }
+    }
+
+    pub fn verify_with_pk(&self, pk: &BlsPublicKey) -> bool {
+        let to_verify = self.key.as_event();
+        match serialisation::serialise(&to_verify) {
+            Ok(data) => pk.verify(&self.sig, data),
+            _ => false,
+        }
+    }
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct SectionProofChain {
+    genesis_pk: BlsPublicKey,
+    blocks: Vec<SectionProofBlock>,
+}
+
+impl SectionProofChain {
+    pub fn from_genesis(pk: BlsPublicKey) -> Self {
+        Self {
+            genesis_pk: pk,
+            blocks: Vec::new(),
+        }
+    }
+
+    pub fn push(&mut self, block: SectionProofBlock) {
+        self.blocks.push(block);
+    }
+
+    #[allow(unused)]
+    pub fn validate(&self) -> bool {
+        let mut current_pk = &self.genesis_pk;
+        for block in &self.blocks {
+            if !block.verify_with_pk(current_pk) {
+                return false;
```

Comment:
What kind of message would you have in mind, though? The blocks don't really contain any meaningful data that would make it easier to identify the offender - it's a public key and a signature, neither is very readable.

Reply:
maybe the public key (which is a public id)?
In case of failure, we need to find out who is failing anyway, right?

Reply:
The public key here is more of a set of public IDs, not a single one - and it actually contains a full section info... But it makes sense :+1:

Reply:
right, then maybe that shall be inside the `PublicKey::verify`, we shall log which one is failing or whether it's not reaching the threshold.
Maybe no need to log whehter reaching the threshold.
However failing a signature is some behaviour shall be noticed.

Reply:
Would not spend too much time on that, when we debug the first issue we will find out what log we need and we can add it then (it might be neither of the options, as I think we will probably want the id of the current node).

So I'd say, add something minimal if that make sense otherwise happy to leave it as is.

Reply:
I think I'd leave it as is for now, then - adding some logging to the verification with a single public ID would make the code there quite a bit uglier, which we can probably do temporarily when debugging, but I'm not sure if it's worth it to have such code by default.